ucm2: sof-glkda7219max: add HDMI/DP support

plbossart opened the pull request:
First add HDMI/DP

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

plbossart commented:
@perexg can you remind me how we can test those UCM configs now? the ucm validation doesn't seem to have been udpated in a very long time, and when I try with alsacum nothing gets loaded:

````
root@plb-Bobba:~# aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: sofglkda7219max [sof-glkda7219max], device 0: Speakers (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofglkda7219max [sof-glkda7219max], device 1: Headset (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofglkda7219max [sof-glkda7219max], device 5: HDMI1 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofglkda7219max [sof-glkda7219max], device 6: HDMI2 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofglkda7219max [sof-glkda7219max], device 7: HDMI3 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
root@plb-Bobba:~# alsaucm listcards
  list is empty
root@plb-Bobba:~# alsaucm reload
ALSA lib main.c:1412:(snd_use_case_mgr_open) error: failed to import hw:0 use case configuration -2
alsaucm: error failed to open sound card hw:0: No such file or directory


````

plbossart commented:
This UCM issue could also be a consequence of the machine driver renaming, I see this in PulseAudio

````
D: [pulseaudio] module-udev-detect.c: /devices/pci0000:00/0000:00:0e.0/glk_da7219_mx98357a/sound/card0 is busy: no
D: [pulseaudio] module-udev-detect.c: Loading module-alsa-card with arguments 'device_id="0" name="pci-0000_00_0e.0-platform-glk_da7219_mx98357a" card_name="alsa_card.pci-0000_00_0e.0-platform-glk_da7219_mx98357a" namereg_fail=false tsched=yes fixed_latency_range=no ignore_dB=no deferred_volume=yes use_ucm=yes avoid_resampling=no card_properties="module-udev-detect.discovered=1"'
D: [pulseaudio] dbus-util.c: Successfully connected to D-Bus session bus 3dab3efe16d7a012ab85143b62673252 as :1.1
D: [pulseaudio] reserve-wrap.c: Successfully acquired reservation lock on device 'Audio0'
I: [pulseaudio] (alsa-lib)main.c: error: failed to import hw:0 use case configuration -2
I: [pulseaudio] (alsa-lib)main.c: error: failed to import sof-glkda7219max use case configuration -2
I: [pulseaudio] alsa-ucm.c: UCM not available for card sof-glkda7219max
````

plbossart commented:
or a name length limitation?

````
root@plb-Bobba:/proc/asound# more cards
 0 [sofglkda7219max]: sof-glkda7219ma - sof-glkda7219max
                      Google-Bobba-rev2
````

plbossart commented:
Update: the UCM files are now loaded (thanks @perexg for the driver/card name hint), speakers, headphone and headset mic work.
problem areas:
DMIC doesn't seem to work, but this could be because of the 4ch capture. We need the ability to extract 2ch as done by CRAS with
    
    CaptureChannelMap "0 0 -1 -1 -1 -1 -1 -1 -1 -1 -1"
    
Automatic detection of headset/headphone doesn't seem to work, all devices show as 'n/a - Celeron/Pentium Silver Processor High Definition Audio'. It's probably because of ALSA mixer issues or the use of these definitions:

            CaptureMixerElem "Headset Mic"
            JackDev "sofglkda7219max Headset Jack"

this last value is problematic anyways since we also need to test on APL, not just GLK.

perexg commented:
> Update: the UCM files are now loaded (thanks @perexg for the driver/card name hint), speakers, headphone and headset mic work. problem areas: DMIC doesn't seem to work, but this could be because of the 4ch capture. We need the ability to extract 2ch as done by CRAS with
> 
> ```
> CaptureChannelMap "0 0 -1 -1 -1 -1 -1 -1 -1 -1 -1"
> ```

The channel split is being implemented in https://github.com/alsa-project/alsa-ucm-conf/issues/154 (my `macro` branch). The indexes are exposed as `[Direction]ChannelX` (index to the hw stream) and `[Direction]ChannelPosX` (sound position like FR = front right) values for applications or (by default) the alsa-lib plugins are used to completely hide the split in the UCM / alsa-lib.

Example config: https://github.com/perexg/alsa-ucm-conf/blob/macro/ucm2/USB-Audio/Behringer/UMC204HD-HiFi.conf

> Automatic detection of headset/headphone doesn't seem to work, all devices show as 'n/a - Celeron/Pentium Silver Processor High Definition Audio'. It's probably because of ALSA mixer issues or the use of these definitions:
> 
> ```
>         CaptureMixerElem "Headset Mic"
>         JackDev "sofglkda7219max Headset Jack"
> ```
> 
> this last value is problematic anyways since we also need to test on APL, not just GLK.

Add `JackControl` value (assuming that the jack is exposed as the ALSA jack control, too). PA/PW does not handle `JackDev` at the moment.

perexg commented on ucm2/Intel/sof-glkda7219max/HiFi.conf:
```diff
@@ -0,0 +1,89 @@
+	Value {
+		CapturePCM "hw:${CardId},99"
+		CaptureChannels 4
+		#CaptureChannelMap "0 0 -1 -1 -1 -1 -1 -1 -1 -1 -1"
```
Remove CRAS specific configs.

plbossart commented on ucm2/Intel/sof-glkda7219max/HiFi.conf:
```diff
@@ -0,0 +1,89 @@
+	}
+
+	EnableSequence [
+		cset "name='Playback Digital Volume' 111"
```
yes, agreed.

plbossart replied:
done

plbossart commented on ucm2/Intel/sof-glkda7219max/HiFi.conf:
```diff
@@ -0,0 +1,89 @@
+# command-line sequence to switch playback/capture
+# alsaucm -c sof-glkda7219max set _verb HiFi set _enadev Headphone
```
?? that works fine for me:

````
root@plb-Bobba:~# aplay -l
**** List of PLAYBACK Hardware Devices ****
card 0: sofglkda7219max [sof-glkda7219max], device 0: Speakers (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofglkda7219max [sof-glkda7219max], device 1: Headset (*) []
  Subdevices: 0/1
  Subdevice #0: subdevice #0
card 0: sofglkda7219max [sof-glkda7219max], device 5: HDMI1 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofglkda7219max [sof-glkda7219max], device 6: HDMI2 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: sofglkda7219max [sof-glkda7219max], device 7: HDMI3 (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
root@plb-Bobba:~# alsaucm -c sof-glkda7219max set _verb HiFi set _enadev InternalMic
root@plb-Bobba:~# alsaucm -c sof-glkda7219max set _verb HiFi set _enadev Headphone
````

plbossart commented on ucm2/Intel/sof-glkda7219max/HiFi.conf:
```diff
@@ -0,0 +1,89 @@
+	]
+}
+
+SectionDevice."InternalMic".0 {
```
done

plbossart commented:
> Add `JackControl` value (assuming that the jack is exposed as the ALSA jack control, too). PA/PW does not handle `JackDev` at the moment.

I changed the JackControl, no obvious change in the behavior. I still see the 'n/a - Celeron/Pentium Silver Processor High Definition Audio'. I have no idea what this CaptureElemMixer is supposed to be set to...

perexg commented:
> > `EnableSequence` in verb should set mixer/routing to a sane state (off) for all devices before application activates a device. I introduced `disdevall` sequence command in recent alsa-lib (upcoming 1.2.7) to make this easier (Syntax 6) - [alsa-project/alsa-lib@4119e33](https://github.com/alsa-project/alsa-lib/commit/4119e33d280352afc337dd01769f9fc8424c82af) and [15a1250](https://github.com/alsa-project/alsa-ucm-conf/commit/15a12508003c95fd541b62ebb71b87548d2089dc) .
> 
> Makes sense, but to be clear does 'disdevall' actually turn off all switches in the card? if yes, this could potentially affect unrelated switches that are turned on by default, e.g. on the SOF side, and are common to multiple outputs. There are a number of controls that are added in the topology for test/convenience but are expected to remain 'on' at all times.

The command does not touch all switches. It runs the disable sequences for all devices in the verb. So it's configuration controlled. If you need a specific order or only a set of sequences, you can use new command `devdis2` which runs only sequence for one device (like `disdev2 "Mic2"`).

plbossart commented:
> Is the 'Headphone Jack' control set to true (something is detected in the jack) ?

@perexg The "Headset Jack" control does detect a change when inserting the plug:

root@plb-Bobba:~# amixer -Dhw:0 cget numid=65
numid=65,iface=CARD,name='Headset Jack'
  ; type=BOOLEAN,access=r-------,values=1
  : values=off
root@plb-Bobba:~# amixer -Dhw:0 cget numid=65
numid=65,iface=CARD,name='Headset Jack'
  ; type=BOOLEAN,access=r-------,values=1
  : values=on

However PulseAudio/Gnome doesn't toggle automatically. That may be a Ubuntu/userspace issue, for some reason I am unable to reinstall Fedora on the test Chromebook.

I fixed the 'n/a' in the profile, I didn't realize it used the Comment string, fixed now.

perexg commented:
I pushed the initial version e8e8e9d471ef62359228d23b497c5cdd9a3d513f to the ucm config repo with the latest hdmi changes.

TODO:

- [ ] 2ch split for Mic
- [ ] add device disable sequence to verb

WeirdTreeThing commented:
@plbossart @perexg I've added dmic split into my [ucm](https://github.com/WeirdTreeThing/chromebook-ucm-conf/blob/main/glk/sof-glkda7219ma/HiFi.conf) for this card. I'm not sure if/how I should make this into a macro for other chromebooks though.

plbossart commented:
closing, I no longer have a working device to test this.
